File: algorithm/climbing-stairs/climbing-stairs.cpp
class Solution {
public:
    int climbStairs(int n) {
        int x = 1, y = 1;
        for (int i = 0; i < n; i++) {
            int t = y;
            y += x;
            x = t;
        }
        return x;
    }
};

File: algorithm/climbing-stairs/climbing-stairs.go
package climbing_stairs

func climbStairs(n int) int {
	cache := map[int]int{
		0: 0,
		1: 1,
		2: 2,
	}
	return doClimbStairs(n, cache)
}

func doClimbStairs(n int, cache map[int]int) int {
	if v, ok := cache[n]; ok {
		return v
	}

	res := doClimbStairs(n-1, cache) + doClimbStairs(n-2, cache)
	cache[n] = res
	return res
}
